skills/repo.go

```diff
@@ -19,10 +19,201 @@
+
+	return r.unlockedType(typeName)
+}
+
+// AddType adds a skill type to the repository.
+// NOTE: API exception, Type is an interface, so it cannot use simple types as arguments.
+func (r *Repository) AddType(t Type) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	typeName := t.Name()
+	if err := ValidateName(typeName); err != nil {
+		return err
+	}
+	if i, found := r.unlockedTypeIndex(typeName); !found {
+		r.types = append(r.types[:i], append([]Type{t}, r.types[i:]...)...)
+		return nil
+	}
+	return ErrDuplicate
+}
+
+// AllSkills returns all skills of the given type.
+// If skillType is the empty string, all skills are returned.
+func (r *Repository) AllSkills(skillType string) []*Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Skill
+	if skillType == "" {
+		result = make([]*Skill, len(r.skills))
+		copy(result, r.skills)
+	} else {
+		result = make([]*Skill, 0)
+		for _, skill := range r.skills {
+			if skill.Type == skillType {
+				result = append(result, skill)
+			}
+		}
+	}
+	return result
+}
+
+// Skills returns the skills offered by the named snap.
+func (r *Repository) Skills(snapName string) []*Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Skill
+	for _, skill := range r.skills {
+		if skill.Snap == snapName {
+			result = append(result, skill)
+		}
+	}
+	return result
+}
+
+// Skill returns the specified skill from the named snap.
+func (r *Repository) Skill(snapName, skillName string) *Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.unlockedSkill(snapName, skillName)
+}
+
+// AddSkill adds a skill to the repository.
+// Skill names must be valid snap names, as defined by ValidateName.
+// Skill name must be unique within a particular snap.
+func (r *Repository) AddSkill(snapName, skillName, typeName, label string, attrs map[string]interface{}) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Reject skill with invalid names
+	if err := ValidateName(snapName); err != nil {
+		return err
+	}
+	if err := ValidateName(skillName); err != nil {
+		return err
+	}
+	// TODO: ensure that given snap really exists
+	t := r.unlockedType(typeName)
+	if t == nil {
+		return ErrTypeNotFound
+	}
+	skill := &Skill{
+		Name:  skillName,
+		Snap:  snapName,
+		Type:  typeName,
+		Attrs: attrs,
+		Label: label,
+	}
+	// Reject skill that don't pass type-specific sanitization
+	if err := t.Sanitize(skill); err != nil {
+		return err
+	}
+	if i, found := r.unlockedSkillIndex(snapName, skillName); !found {
+		r.skills = append(r.skills[:i], append([]*Skill{skill}, r.skills[i:]...)...)
+		return nil
+	}
+	return ErrDuplicate
+}
+
+// RemoveSkill removes the named skill provided by a given snap.
+// Removing a skill that doesn't exist returns a ErrSkillNotFound.
+// Removing a skill that is granted returns ErrSkillBusy.
+func (r *Repository) RemoveSkill(snapName, skillName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// TODO: Ensure that the skill is not used anywhere
+	for i, skill := range r.skills {
```

Comment:
That can actually be a trivial map operation, as suggested. Let's please do it in this branch.